common/flatpak-dir.c

```diff
@@ -4941,26 +4941,43 @@ flatpak_dir_setup_extra_data (FlatpakDir                           *self,
   /* If @results is set, @rev must be. */
   g_assert (results == NULL || rev != NULL);
 
-  extra_data_sources = flatpak_repo_get_extra_data_sources (repo, rev, cancellable, NULL);
-  if (extra_data_sources == NULL)
+  /* ostree-metadata and appstreams never have extra data, so ignore those */
+  if (g_str_has_prefix ("app/", ref) || g_str_has_prefix ("runtime/", ref))
     {
-      /* Pull the commits (and only the commits) to check for extra data
-       * again. Here we don't pass the progress because we don't want any
-       * reports coming out of it. */
-      if (!repo_pull (repo, repository,
-                      NULL,
-                      ref,
-                      rev,
-                      token,
-                      results,
-                      flatpak_flags,
-                      OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY,
-                      NULL,
-                      cancellable,
-                      error))
-        return FALSE;
-
       extra_data_sources = flatpak_repo_get_extra_data_sources (repo, rev, cancellable, NULL);
+      if (extra_data_sources == NULL)
+        {
+          /* This is a gigantic hack where we download the commit in a temporary transaction
+           * which we then abort after having read the result. We do this to avoid creating
+           * a partial commit in the local repo and a ref that points to it, because that
+           * causes ostree to not use static deltas.
+           * See https://github.com/flatpak/flatpak/issues/3412 for details.
+           */
+
+          if (!ostree_repo_prepare_transaction (repo, NULL, cancellable, error))
+            return FALSE;
+
+          /* Pull the commits (and only the commits) to check for extra data
+           * again. Here we don't pass the progress because we don't want any
+           * reports coming out of it. */
+          if (!repo_pull (repo, repository,
+                          NULL,
+                          ref,
+                          rev,
+                          token,
+                          results,
+                          flatpak_flags,
+                          OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY,
+                          NULL,
+                          cancellable,
+                          error))
+            return FALSE;
+
+          extra_data_sources = flatpak_repo_get_extra_data_sources (repo, rev, cancellable, NULL);
+
+          if (!ostree_repo_abort_transaction (repo, cancellable, error))
+            return FALSE;
```

Comment:
Instead of this dance with aborting the transaction, can you pull just the rev without the ref? That's supported in ostree and then you won't update the ref and you'll avoid the issue where ostree sees a partial rev on the ref when the local ref didn't exist before.

Reply:
Does pulling a rev without a ref work with the p2p case? I was sorta assuming it didn't. Yet another reason why p2p should die in flames.